### ISSUE TITLE
whether batch merge into is supported

### ISSUE THREAD

**Issue by freeNestor** [sijms/go-ora#343]
I want to migrate godror to go-ora because cross-compile is hard. But I encountered a problem after changed to go-ora.
Our app batch writes data with merge into sql with no error when using godror, but batch writes fail with error below when using go-ora. I want to known is there a batch execute sql method with go-ora, or is it support merge into sql?
`bulk merge into error: ORA-01484: arrays can only be bound to PL/SQL statements`

batch code snippet:
`_, err = getconn.ExecContext(ctxTimeout, sqlbufstr, bindslice...)`
bindslice is like [][]string, [0][]string is for first bind parameter batched, [1][]string for second, etc.

how to do this with go-ora?
Thx.

**Comment by freeNestor** [comment]
> would you please send example sql text to test it on my own environment

It's long so I use '...' for short, thx.

`MERGE INTO schema.tablename t1 USING(select :1 COL1,:2 COL2,:3 COL3,:4 COL4 from dual) tmp 
ON (tmp.COL1=t1.COL1 AND tmp.COL2=t1.COL2 AND tmp.COL3=t1.COL3 AND tmp.COL4=t1.COL4) 
WHEN MATCHED THEN UPDATE SET colx=:5,coly=:6,...,coln=:118 
WHERE t1.COL1=:119 AND t1.COL2=:120 AND t1.COL3=:121 AND t1.COL4=:122 AND datecol <= TO_TIMESTAMP(:123,'YYYY-MM-DD HH24:MI:SS.FF6') 
WHEN NOT MATCHED THEN INSERT (COL1,COL2,COL3,COL4,...,datecol) VALUES (:124,:125,:126,:127,...,:241)`

**Comment by sijms** [comment]
does your parameter is of type array or simple types like int64, float64, time.Time and string?
if your type is array so just pass an array so use array like []string, []int64, float64 and []time.Time

the default behavior is translate parameter according to its position
if you need to use named parameter so use sql.Named("name", value)

**Comment by sijms** [comment]
another thing that may describe your code
you need to execute multiple rows so instead of make like this
```golang
for x:=0; x < len(rows); x++ {
    _, err = conn.ExecContext(context, sqlText, rows[x])
}
```
you need to pass context, sqlText and rows (as array) and make one call of exec

I do this with insert in function BulkInsert but not with merge 
if this is the situation so I will add new function BulkMerge

**Comment by freeNestor** [comment]
> > does your parameter is of type array or simple types like int64, float64, time.Time and string? if your type is array so just pass an array so use array like []string, []int64, float64 and []time.Time
> > the default behavior is translate parameter according to its position if you need to use named parameter so use sql.Named("name", value)
> 
> I don't know how to describe this. I do execute one call, not loop. I dont use named parameter. My parameter 'bindslice' is array of interface{} ([]interface{}), and all of underlying elements are []string. here is a snapshot for godror batch execute doc may give more info.
> 
> <img alt="image" width="1084" src="https://user-images.githubusercontent.com/19606302/231324209-5e3311ed-c951-41e2-ae45-a13fe39d31e3.png">

Though this doc for insert, but merge into is also work.

**Comment by sijms** [comment]
ok now I get your point
I have a function in the 
```golang
func (conn *Connection) BulkInsert(sqlText string, rowNum int, columns ...[]driver.Value) (*QueryResult, error)
```
I think it will be the same
you can see `examples/bulk_insert` for code example
would you please test it and see if it will work?

**Comment by freeNestor** [comment]
> ok now I get your point I have a function like this
> 
> ```go
> func (conn *Connection) BulkInsert(sqlText string, rowNum int, columns ...[]driver.Value) (*QueryResult, error)
> ```
> 
> I think it will be the same you can see `examples/bulk_insert` for code example would you please test it and see if it will work?

I've checked this, so I must use *Connection interface not *sql.DB nor *sql.Conn. I'll test it, but I appreciate if I could use *sql.DB or *sql.Conn.

**Comment by freeNestor** [comment]
I just tested it fails.

<img width="865" alt="image" src="https://user-images.githubusercontent.com/19606302/231341640-51c91247-0b51-42d1-bffc-835fb181f2af.png">


The first error occur:
`bulk merge into error: read tcp 192.168.x.x:50782->192.168.x.x:1521: read: connection reset by peer`

 error after first:
`bulk merge into error: EOF`
`bulk merge into error: write tcp 192.168.x.x:50782->192.168.x.x:1521: write: broken pipe`

and soon panic:

[2023-04-12T11:21:08.283] [FATAL] [data/global.go:2509] goroutine 167 [running]:
runtime/debug.Stack()
        /usr/local/go/src/runtime/debug/stack.go:24 +0x65
octopus2.0/pkg/data.PanicStackTrace()
        /Users/engfall/Documents/90-Develop/workspace/octopus/test/pkg/data/global.go:2509 +0x4b
panic({0xbc1e60, 0xc00142d9e0})
        /usr/local/go/src/runtime/panic.go:884 +0x213
bytes.(*Buffer).Write(0xc0022101c0?, {0xc0020016ba?, 0xc0020016d0?, 0xc002001602?})
        /usr/local/go/src/bytes/buffer.go:172 +0xd6
github.com/sijms/go-ora/v2/network.(*Session).PutBytes(...)
        /Users/engfall/Documents/90-Develop/go-src/pkg/mod/github.com/sijms/go-ora/v2@v2.6.15/network/session.go:841
github.com/sijms/go-ora/v2.(*ParameterInfo).write(0xc002001858, 0xc0022101c0)
        /Users/engfall/Documents/90-Develop/go-src/pkg/mod/github.com/sijms/go-ora/v2@v2.6.15/parameter.go:304 +0x27a
github.com/sijms/go-ora/v2.(*defaultStmt).basicWrite(0xc0001e21e0, 0x88029, 0x1, 0x0)
        /Users/engfall/Documents/90-Develop/go-src/pkg/mod/github.com/sijms/go-ora/v2@v2.6.15/command.go:226 +0x985
github.com/sijms/go-ora/v2.(*Connection).BulkInsert(0xc000002000, {0xc001ef0000?, 0x4?}, 0x29, {0xc0025c5800, 0xf1, 0xc002001cf0?})
        /Users/engfall/Documents/90-Develop/go-src/pkg/mod/github.com/sijms/go-ora/v2@v2.6.15/connection.go:1059 +0x745
octopus2.0/pkg/data.(*OraSource).BatchWrite(0xc00020a740, 0xc00016d230?, 0x4?, 0xc105a6c9042e6d1b?, {0xc0001b8800, 0x29, 0x0?}, 0x0?)
        /Users/engfall/Documents/90-Develop/workspace/octopus/test/pkg/data/oracle.go:356 +0xb74
octopus2.0/pkg/data.TaskWriteDBv2.func1({0xd54210, 0xc0000ea5a0}, 0xc000484c60?, 0x4, 0x0, 0xc0000cc750)
        /Users/engfall/Documents/90-Develop/workspace/octopus/test/pkg/data/global.go:1304 +0x1bbf
created by octopus2.0/pkg/data.TaskWriteDBv2
        /Users/engfall/Documents/90-Develop/workspace/octopus/test/pkg/data/global.go:1123 +0x488
 {"error": "runtime error: slice bounds out of range [7650:7648]"}

**Comment by sijms** [comment]
fixed in v2.6.17

**Comment by sijms** [comment]
you can see example code in `examples/merge` I also plan to make support for insert struct with tag and its array 
so instead of passing **array1, array2, array3** as parameters you can pass one struct array and each field will translated into  **array1, array2, array3**

**Comment by freeNestor** [comment]
> repetition
@sijms Thx very much. I'll test this.

**Comment by freeNestor** [comment]
@sijms 
Sorry, I tested failed. Am I using it wrong? pls help.
Because columns are stored in a variable, I shouldn't hard code them, so I write for loop to iterate. And more, I convert all datas to string because data type cannot be determined before using.
Code like below:
`bindnew := make([]interface{},0)`
...
`for colname := range cols {`
`    var data []string`
`    for ... { `
`         some convert codes`
`         data = append(data, val)`
`         ...`
`    }`
`     ...`
`    bindnew = append(bindnew, sql.Named(colname, data))`
`}`
`...`
`_, err = getconn.ExecContext(ctxTimeout, sqlbufstr, bindnew...)`

error:  bulk merge into error: ORA-24381: error(s) in array DML\n

**Comment by sijms** [comment]
this error means that there is something wrong passed to server. you can see this in log file just append `trace file=trace.log` to the connection url and see the content of the log file

**Comment by freeNestor** [comment]
> this error means that there is something wrong passed to server. you can see this in log file just append `trace file=trace.log` to the connection url and see the content of the log file

I add trace log, when writing only one record the error is "bulk merge into error: ORA-01843: not a valid month\n", when writing multiple records the error is "ORA-24381: error(s) in array DML\n". I think the problem is timestamp data not being recognized by database. Trace log file gives me no more useful info to analyze.
My table has timestamp type columns, data string to be writing like:
`"2022-12-07T09:05:16.648449Z"`
or
`"2022-12-07T09:05:16.648Z"`
just fraction is different.
Could you pls make sure that timestamp is supported for writing with this batch mode? Thx.

**Comment by sijms** [comment]
to pass timestamp to the driver you should use `go_ora.TimeStamp` or `go_ora.TimeStampTZ` 
if you pass `time.Time` it will be interpreted as `Date`
string is not accepted as timestamp parameter
if you need to use string so you should use something like pl/sql `TO_TIMESTAMP`
I am now working on upgrading the driver to use go struct as a parameter and  benefit from struct tag
for example
```golang
type MyData struct {
	id         int    `db:"ID"`
	name       string `db:"NAME"`
	floatValue string `db:"FVAL,float"`
	dateValue  string `db:"DATE,date"`
	timeStamp  string `db:"TS,timestamp"`
}
```
as you see you can specify name, type,size,direction in the tag so the operation is not dependant on field type

**Comment by freeNestor** [comment]
> to pass timestamp to the driver you should use `go_ora.TimeStamp` or `go_ora.TimeStampTZ` if you pass `time.Time` it will be interpreted as `Date` string is not accepted as timestamp parameter if you need to use string so you should use something like pl/sql `TO_TIMESTAMP` I am now working on upgrading the driver to use go struct as a parameter and benefit from struct tag for example
> 
> ```go
> type MyData struct {
> 	id         int    `db:"ID"`
> 	name       string `db:"NAME"`
> 	floatValue string `db:"FVAL,float"`
> 	dateValue  string `db:"DATE,date"`
> 	timeStamp  string `db:"TS,timestamp"`
> }
> ```
> 
> as you see you can specify `name, type,size,direction` in the tag so the operation is not dependant on field type

Sorry, test failed again😰, same error "bulk merge into error: ORA-01843: not a valid month\n"

Code:
```
...
                var bind []interface{}
		for j := range datas {
			val := datas[j][colname]
			if coltype == "TIMESTAMP" || coltype == "DATETIME" {
				if val != nil {
					valt, err = TryParseDateTime(val.(string), timeloc)
					if err != nil {
						return err
					}
					bind = append(bind, go_ora.TimeStamp(valt))
				} else {
					bind = append(bind, go_ora.NullTimeStamp{Valid: false})
				}
			} else {
				datatodb, err := ExtractJsontoString(val, "oracle")
				if err != nil {
					return err
				}
				bind = append(bind, datatodb)
			}
		}
		// bindtailslice[i] = bind
		bindnew = append(bindnew, sql.Named(colname, bind))
...```

**Comment by freeNestor** [comment]
> to pass timestamp to the driver you should use `go_ora.TimeStamp` or `go_ora.TimeStampTZ` if you pass `time.Time` it will be interpreted as `Date` string is not accepted as timestamp parameter if you need to use string so you should use something like pl/sql `TO_TIMESTAMP` I am now working on upgrading the driver to use go struct as a parameter and benefit from struct tag for example
> 
> ```go
> type MyData struct {
> 	id         int    `db:"ID"`
> 	name       string `db:"NAME"`
> 	floatValue string `db:"FVAL,float"`
> 	dateValue  string `db:"DATE,date"`
> 	timeStamp  string `db:"TS,timestamp"`
> }
> ```
> 
> as you see you can specify `name, type,size,direction` in the tag so the operation is not dependant on field type

Thx again. Solved timestamp problem, but another problem occurred😰. Writing successfully but database record count is inconsistent, some records are missing.
```
[2023-04-23T14:28:39.923] [INFO] [data/global.go:1314] flow-batchwrite  {"taskname": "xx", "driver": "oracle", "flowid": 4, "goroutine NO.": 0, "write-count": 1, "duration(ms)": 38}
[2023-04-23T14:28:40.608] [INFO] [data/global.go:1433] flow-batchwrite  {"taskname": "xx", "driver": "oracle", "flowid": 4, "goroutine NO.": 0, "write-count": 156, "duration(ms)": 79}
```
database select count(1) show only 83 count. Trace log shows no error, app shows no error returned.

**Comment by sijms** [comment]
I think this is from merge itself because if you use bulk insert all your record will be inserted 
so merge sql should be reviewed carefully to see if there are some record updated
also I face same problem when I create my examples because some values is null and when you test null and "" for equality it will give unequal

**Comment by freeNestor** [comment]
No, it behaves odd with go_ora bulk merge.My scenario is select from a source table, then merge into another table.Using merge you must supply primary key or unique key to match, because insert into target table may cause unique constraint without merge, so if matched will update, if not matched will insert, other fields value cannot change matched result if null or "", you should not use other key than primary key with merge sql. I can confirm that these 157 count data's primary keys are unique, so update should never happened in test this time , but test failure with go_ora driver, it should be 157 count records in target table.  It behaves right with godror compared finally.

**Comment by freeNestor** [comment]
> ok can you give me example code to produce the error and investigate

example table definition：
```
CREATE TABLE TESTSHORT 
   (	"ID" NUMBER(20,0) NOT NULL ENABLE, 
	"TM" VARCHAR2(30), 
	"SN" VARCHAR2(25) NOT NULL ENABLE, 
	"CUS" VARCHAR2(20) NOT NULL ENABLE, 
	"AID" VARCHAR2(20), 
	"TR" VARCHAR2(8) NOT NULL ENABLE, 
	"PID" VARCHAR2(20), 
	"CODE" VARCHAR2(20) NOT NULL ENABLE, 
	"TTNO" VARCHAR2(20), 
	"UPDATETIME" TIMESTAMP (6) DEFAULT systimestamp, 
	 PRIMARY KEY ("ID")
)
```

MERGE SQL:

```
MERGE INTO testdbuser.testshort t1 USING(select :ID ID from dual) tmp ON (tmp.ID=t1.ID) WHEN MATCHED THEN UPDATE SET TM=:TM,SN=:SN,CUS=:CUS,AID=:A
PPID,TR=:TR,PID=:PID,CODE=:CODE,TTNO=:TTNO,UPDATETIME=:UPDATETIME WHERE t1.ID=:ID AND t1.UPDATETIME<=:UPDATETIME WHEN NOT MATCHED THEN INSERT (ID,TM,T
RXSN,CUS,AID,TR,PID,CODE,TTNO,UPDATETIME) VALUES (:ID,:TM,:SN,:CUS,:AID,:TR,:PID,:CODE,:TTNO,:UPDATETIME)
```

bulk merge code:
```
func (ora *OraSource) BatchWrite(fid int, pool *sqlx.DB, dsttbl *TableObject, datas []interface{}, bindnum int) error {
	bindnew := make([]interface{}, 0)
	var colname, coltype string
	var err error
	var valt time.Time
	for _, colinfo := range dstblcoltypes[fid]["cols"].([]map[string]string) {
		colname = colinfo["name"]
		coltype = colinfo["type"]

		var bind []interface{}
		for _, valmap := range datas {
			val := valmap.(map[string]interface{})[colname]
			if coltype == "TIMESTAMP" || coltype == "DATETIME" ||
				coltype == "TimeStampDTY" {
				if val != nil {
					valt, err = TryParseDateTime(val.(string), timeloc)
					if err != nil {
						return err
					}
					bind = append(bind, go_ora.TimeStamp(valt))
				} else {
					bind = append(bind, go_ora.NullTimeStamp{Valid: false})
				}
			} else {
				bind = append(bind, val)
			}
		}
		bindnew = append(bindnew, sql.Named(colname, bind))
	}
	sqlbufstr := flowmergesql[fid]
	ctxTimeout, cancel := context.WithTimeout(context.Background(), time.Second*time.Duration(WriteTimeout))
	defer cancel()
	getconn, err := pool.Connx(ctxTimeout)
	if err != nil {
		return err
	}
	defer getconn.Close()
	stmt, err := getconn.PrepareContext(ctxTimeout, sqlbufstr)
	if err != nil {
		return err
	}
	defer stmt.Close()
	_, err = stmt.ExecContext(ctxTimeout, bindnew...)
	if err != nil {
		return fmt.Errorf("bulk merge into error: %v - sql: %s", err, sqlbufstr)
	}
	return nil
}
```

In test make source table TM and TTNO column data null, other has value, ID is primary key and unique. make source table more than 157 count records and you can go.

**Comment by sijms** [comment]
this is the code I test
```golang
package main

import (
	"database/sql"
	"fmt"
	_ "github.com/sijms/go-ora/v2"
	go_ora "github.com/sijms/go-ora/v2"
	"os"
	"time"
)

func createTable(conn *sql.DB) error {
	t := time.Now()
	sqlText := `CREATE TABLE TESTSHORT 
   (	"ID" NUMBER(20,0) NOT NULL ENABLE, 
	"TM" VARCHAR2(30), 
	"SN" VARCHAR2(25) NOT NULL ENABLE, 
	"CUS" VARCHAR2(20) NOT NULL ENABLE, 
	"AID" VARCHAR2(20), 
	"TR" VARCHAR2(8) NOT NULL ENABLE, 
	"PID" VARCHAR2(20), 
	"CODE" VARCHAR2(20) NOT NULL ENABLE, 
	"TTNO" VARCHAR2(20), 
	"UPDATETIME" TIMESTAMP (6) DEFAULT systimestamp, 
	 PRIMARY KEY ("ID")
)`
	_, err := conn.Exec(sqlText)
	if err != nil {
		return err
	}
	fmt.Println("finish create table: ", time.Now().Sub(t))
	return nil
}

func dropTable(conn *sql.DB) error {
	t := time.Now()
	_, err := conn.Exec("drop table TESTSHORT purge")
	if err != nil {
		return err
	}
	fmt.Println("finish drop table: ", time.Now().Sub(t))
	return nil
}

func merge(conn *sql.DB) error {
	t := time.Now()
	sqlText := `MERGE INTO testshort t1 USING(select :ID ID from dual) tmp ON (tmp.ID=t1.ID) 
    WHEN MATCHED THEN UPDATE SET TM=:TM,SN=:SN,CUS=:CUS, AID=:AID,TR=:TR,PID=:PID,CODE=:CODE,TTNO=:TTNO,UPDATETIME=:UPDATETIME WHERE t1.ID=:ID AND t1.UPDATETIME<=:UPDATETIME 
    WHEN NOT MATCHED THEN INSERT (ID,TM,SN,CUS,AID,TR,PID,CODE,TTNO,UPDATETIME) VALUES (:ID,:TM,:SN,:CUS,:AID,:TR,:PID,:CODE,:TTNO,:UPDATETIME)`
	length := 500
	id := make([]int, length)
	tm := make([]sql.NullString, length)
	sn := make([]string, length)
	cus := make([]string, length)
	aid := make([]sql.NullString, length)
	tr := make([]string, length)
	pid := make([]string, length)
	code := make([]string, length)
	ttno := make([]sql.NullString, length)
	updateTime := make([]go_ora.TimeStamp, length)
	for x := 0; x < length; x++ {
		id[x] = x + 1
		if x > 0 && x%10 == 0 {
			tm[x] = sql.NullString{String: "", Valid: false}
			aid[x] = sql.NullString{String: "", Valid: false}
			ttno[x] = sql.NullString{String: "", Valid: false}
		} else {
			tm[x] = sql.NullString{String: "tm text", Valid: true}
			aid[x] = sql.NullString{String: "aid text", Valid: true}
			ttno[x] = sql.NullString{String: "ttno text", Valid: false}
		}
		sn[x] = "sn text"
		cus[x] = "cus text"
		tr[x] = "tr text"
		pid[x] = "pid text"
		code[x] = "code text"
		updateTime[x] = go_ora.TimeStamp(time.Now())
	}
	_, err := conn.Exec(sqlText, sql.Named("ID", id),
		sql.Named("TM", tm),
		sql.Named("SN", sn),
		sql.Named("CUS", cus),
		sql.Named("AID", aid),
		sql.Named("TR", tr),
		sql.Named("PID", pid),
		sql.Named("CODE", code),
		sql.Named("TTNO", ttno),
		sql.Named("UPDATETIME", updateTime))

	if err != nil {
		return err
	}
	fmt.Println("finish merge: ", time.Now().Sub(t))
	return nil
}
func main() {
	conn, err := sql.Open("oracle", os.Getenv("DSN"))
	if err != nil {
		fmt.Println("can't open connection: ", err)
		return
	}
	defer func() {
		err = conn.Close()
		if err != nil {
			fmt.Println("can't close connection: ", err)
			return
		}
	}()
	err = createTable(conn)
	if err != nil {
		fmt.Println("can't create table: ", err)
		return
	}
	//defer func() {
	//	err = dropTable(conn)
	//	if err != nil {
	//		fmt.Println("can't drop table: ", err)
	//	}
	//}()
	err = merge(conn)
	if err != nil {
		fmt.Println("can't merge: ", err)
		return
	}
}
```
after run code
```sql
SELECT COUNT(*) FROM TESTSHORT
```
result
```bash
500
```

**Comment by freeNestor** [comment]
Thx very much for your help. I modified your code a little and test, it's also normal. I test myself app code and investigated data deeply, found it relates gjson, it wrongly unmarshalled two fields json value. I'll fix it and continue test go_ora.
Thx again.

**Comment by sijms** [comment]
structure pars pre-release v2.7.1
you can test also

**Comment by sijms** [comment]
you will find an example in `examples/struct pars/main.go`

**Comment by freeNestor** [comment]
@sijms 
We got another error when using bulk merge. What does this mean?
```
"error": "bulk merge into error: struct parameter only allowed with user defined type (UDT)
```

**Comment by freeNestor** [comment]
Does bulk merge support context timeout? Testing below code results main stuck long time and not expected behavior.
Code below simulates primary key conflict to see if it behaves right.
```
package main

import (
	"context"
	"database/sql"
	"fmt"
	"math/rand"
	"strconv"
	"sync"
	"time"

	"github.com/jmoiron/sqlx"
	_ "github.com/sijms/go-ora/v2"
	go_ora "github.com/sijms/go-ora/v2"
)

func createTable(conn *sql.DB) error {
	t := time.Now()
	sqlText := `CREATE TABLE TESTSHORT 
   (	"ID" NUMBER(20,0) NOT NULL ENABLE, 
	"TM" VARCHAR2(30), 
	"SN" VARCHAR2(25) NOT NULL ENABLE, 
	"CUS" VARCHAR2(20) NOT NULL ENABLE, 
	"AID" VARCHAR2(20), 
	"TR" VARCHAR2(8) NOT NULL ENABLE, 
	"PID" VARCHAR2(20), 
	"CODE" VARCHAR2(20) NOT NULL ENABLE, 
	"TTNO" VARCHAR2(20), 
	"UPDATETIME" TIMESTAMP (6) DEFAULT systimestamp, 
	"OUT_TEST" VARCHAR2(200), 
	 PRIMARY KEY ("ID")
)`
	_, err := conn.Exec(sqlText)
	if err != nil {
		return err
	}
	fmt.Println("finish create table: ", time.Now().Sub(t))
	return nil
}

func dropTable(conn *sql.DB) error {
	t := time.Now()
	_, err := conn.Exec("drop table TESTSHORT purge")
	if err != nil {
		return err
	}
	fmt.Println("finish drop table: ", time.Now().Sub(t))
	return nil
}

func merge(conn *sqlx.DB) error {
	t := time.Now()
	sqlText := `MERGE INTO testshort1 t1 USING(select :ID ID from dual) tmp ON (tmp.ID=t1.ID) 
    WHEN MATCHED THEN UPDATE SET TM=:TM,SN=:SN,CUS=:CUS, AID=:AID,TR=:TR,PID=:PID,CODE=:CODE,TTNO=:TTNO,UPDATETIME=:UPDATETIME,OUT_TEST=:OUT_TEST WHERE t1.ID=:ID AND t1.UPDATETIME<=:UPDATETIME 
    WHEN NOT MATCHED THEN INSERT (ID,TM,SN,CUS,AID,TR,PID,CODE,TTNO,UPDATETIME,OUT_TEST) VALUES (:ID,:TM,:SN,:CUS,:AID,:TR,:PID,:CODE,:TTNO,:UPDATETIME,:OUT_TEST)`
	length := 500
	var bindall []interface{}
	testjson := []byte(`{"ag":"a3077220"}`)
	rand.Seed(time.Now().UnixMilli())
	for _, colname := range []string{"ID", "TM", "SN", "CUS", "AID", "TR", "PID", "CODE", "TTNO", "UPDATETIME", "OUT_TEST"} {
		var bind []interface{}
		for x := 0; x < length; x++ {
			switch colname {
			case "ID":
				bind = append(bind, rand.Intn(500)+1)
			case "UPDATETIME":
				bind = append(bind, go_ora.TimeStamp(time.Now()))
			case "TM":
				bind = append(bind, string(testjson))
			default:
				bind = append(bind, "text")
			}

		}
		bindall = append(bindall, sql.Named(colname, bind))
	}
	ctxTimeout, cancel := context.WithTimeout(context.Background(), time.Second*time.Duration(5))
	defer cancel()
	getconn, err := conn.Connx(ctxTimeout)
	if err != nil {
		return err
	}
	defer getconn.Close()
	_, err = getconn.ExecContext(ctxTimeout, sqlText, bindall...)
	if err != nil {
		return err
	}
	fmt.Println("finish merge: ", time.Now().Sub(t))
	return nil
}
func main() {
	urlOpts := map[string]string{
		"CONNECTION TIMEOUT": strconv.Itoa(5),
		// "RETRY_COUNT":        strconv.Itoa(3),
		"PREFETCH_ROWS": "1000",
	}
	connstr := go_ora.BuildUrl(..., urlOpts)
	conn, err := sqlx.Connect("oracle", connstr)
	if err != nil {
		fmt.Println("can't open connection: ", err)
		return
	}
	defer func() {
		err = conn.Close()
		if err != nil {
			fmt.Println("can't close connection: ", err)
			return
		}
	}()
	// err = createTable(conn)
	// if err != nil {
	// 	fmt.Println("can't create table: ", err)
	// 	return
	// }
	//defer func() {
	//	err = dropTable(conn)
	//	if err != nil {
	//		fmt.Println("can't drop table: ", err)
	//	}
	//}()
	st := time.Now()
	var wg sync.WaitGroup
	for i := 0; i < 4; i++ {
		wg.Add(1)
		go func(seq int, wgin *sync.WaitGroup) {
			fmt.Printf("%d: start to merge\n", seq)
			defer wgin.Done()
			err = merge(conn)
			if err != nil {
				fmt.Println("can't merge: ", err)
				return
			}
		}(i, &wg)
	}
	wg.Wait()
	et := time.Now()
	fmt.Println("finished total:", et.Sub(st))
	// time.Sleep(time.Second * 10)
}

```

test result, and truncate table causes error "ORA-00054: resource busy and acquire with NOWAIT specified or timeout expired" before main program exits, this means row locks not released in time. we expect that it will timeouts in 5 seconds and release row locks.

```
# ./test 
3: start to merge
2: start to merge
1: start to merge
0: start to merge
finish merge:  47.98854ms
finish merge:  91.925139ms
can't merge:  ORA-24381: error(s) in array DML

can't merge:  ORA-24381: error(s) in array DML

finished total: 2m9.551480205s
```

**Comment by sijms** [comment]
sorry for late
i am investigating the issue because of goroutines debugging is not easy


when I finish i will inform you

**Comment by sijms** [comment]
after rapid investigation of the issue
I get the following:
1- we have 4 go-routines which is started almost at same time
2- Exec operation is started with session.Write() which will take long time as we send 500 to 1000 items so server will keep
receiving data till finish
3- while server is receiving data another session.Write() commands started.
4- Now the server receive multiple merge requests at same time which make a dead lock inside the server and you can realize that at step 7
5- read operation now need to take place and because of dead lock it will stuck
6- read timeout occurs => break connection  which will make write then read which also will stuck
7- it you wait for time you will receive error server dead lock occur
 
I am trying to return error at step 6

but I think this operation bulk insert/merge should not run in parallel

would you try the code on other drivers and compare the result

**Comment by sijms** [comment]
you can see the trace.log file
```log
00000000  00 00 01 78 06 00 00 00  00 00 66 6f 72 20 72 65  |...x......for re|
00000010  73 6f 75 72 63 65 0a 00  00 01 38 38 4f 52 41 2d  |source....88ORA-|
00000020  30 30 30 36 30 3a 20 64  65 61 64 6c 6f 63 6b 20  |00060: deadlock |
00000030  64 65 74 65 63 74 65 64  20 77 68 69 6c 65 20 77  |detected while w|
00000040  61 69 74 69 6e 67 20 66  6f 72 20 72 65 73 6f 75  |aiting for resou|
00000050  72 63 65 0a 00 00 01 38  38 4f 52 41 2d 30 30 30  |rce....88ORA-000|
00000060  36 30 3a 20 64 65 61 64  6c 6f 63 6b 20 64 65 74  |60: deadlock det|
00000070  65 63 74 65 64 20 77 68  69 6c 65 20 77 61 69 74  |ected while wait|
00000080  69 6e 67 20 66 6f 72 20  72 65 73 6f 75 72 63 65  |ing for resource|
00000090  0a 00 00 01 38 38 4f 52  41 2d 30 30 30 36 30 3a  |....88ORA-00060:|
000000a0  20 64 65 61 64 6c 6f 63  6b 20 64 65 74 65 63 74  | deadlock detect|
000000b0  65 64 20 77 68 69 6c 65  20 77 61 69 74 69 6e 67  |ed while waiting|
000000c0  20 66 6f 72 20 72 65 73  6f 75 72 63 65 0a 00 00  | for resource...|
000000d0  01 38 38 4f 52 41 2d 30  30 30 36 30 3a 20 64 65  |.88ORA-00060: de|
000000e0  61 64 6c 6f 63 6b 20 64  65 74 65 63 74 65 64 20  |adlock detected |
000000f0  77 68 69 6c 65 20 77 61  69 74 69 6e 67 20 66 6f  |while waiting fo|
00000100  72 20 72 65 73 6f 75 72  63 65 0a 00 00 01 38 38  |r resource....88|
00000110  4f 52 41 2d 30 30 30 36  30 3a 20 64 65 61 64 6c  |ORA-00060: deadl|
00000120  6f 63 6b 20 64 65 74 65  63 74 65 64 20 77 68 69  |ock detected whi|
00000130  6c 65 20 77 61 69 74 69  6e 67 20 66 6f 72 20 72  |le waiting for r|
00000140  65 73 6f 75 72 63 65 0a  00 00 02 5f 3d 08 ff ff  |esource...._=...|
00000150  ff ff ff ff fd 60 21 4f  52 41 2d 32 34 33 38 31  |.....`!ORA-24381|
00000160  3a 20 65 72 72 6f 72 28  73 29 20 69 6e 20 61 72  |: error(s) in ar|
00000170  72 61 79 20 44 4d 4c 0a                           |ray DML.|
```

**Comment by freeNestor** [comment]
> after rapid investigation of the issue I get the following: 1- we have 4 go-routines which is started almost at same time 2- Exec operation is started with session.Write() which will take long time as we send 500 to 1000 items so server will keep receiving data till finish 3- while server is receiving data another session.Write() commands started. 4- Now the server receive multiple merge requests at same time which make a dead lock inside the server and you can realize that at step 7 5- read operation now need to take place and because of dead lock it will stuck 6- read timeout occurs => break connection which will make write then read which also will stuck 7- it you wait for time you will receive error server dead lock occur
> 
> I am trying to return error at step 6
> 
> but I think this operation bulk insert/merge should not run in parallel
> 
> would you try the code on other drivers and compare the result

other driver test result, I test several times:
Context timeout occurred at last one time test.

```
# ./test-linux-amd64 
godror WARNING: discrepancy between DBTIMEZONE ("+00:00"=0) and SYSTIMESTAMP ("+08:00"=800) - set connection timezone, see https://github.com/godror/godror/blob/master/doc/timezone.md
3: start to merge
1: start to merge
2: start to merge
0: start to merge
l:  24 bindall:  10
l:  24 bindall:  10
l:  24 bindall:  10
l:  24 bindall:  10
finish merge:  106.468325ms
can't merge:  ORA-00001: unique constraint (TESTDBUSER.SYS_C00265776) violated
can't merge:  ORA-00001: unique constraint (TESTDBUSER.SYS_C00265776) violated
finish merge:  162.112783ms
finished total: 163.075625ms

# ./test-linux-amd64 
godror WARNING: discrepancy between DBTIMEZONE ("+00:00"=0) and SYSTIMESTAMP ("+08:00"=800) - set connection timezone, see https://github.com/godror/godror/blob/master/doc/timezone.md
3: start to merge
0: start to merge
1: start to merge
2: start to merge
l:  24 bindall:  10
l:  24 bindall:  10
l:  24 bindall:  10
l:  24 bindall:  10
finish merge:  45.584669ms
can't merge:  ORA-00060: deadlock detected while waiting for resource
finish merge:  3.968049787s
finish merge:  3.997710481s
finished total: 3.997998878s

# ./test-linux-amd64 
godror WARNING: discrepancy between DBTIMEZONE ("+00:00"=0) and SYSTIMESTAMP ("+08:00"=800) - set connection timezone, see https://github.com/godror/godror/blob/master/doc/timezone.md
3: start to merge
1: start to merge
2: start to merge
0: start to merge
l:  24 bindall:  10
l:  24 bindall:  10
l:  24 bindall:  10
l:  24 bindall:  10
finish merge:  40.336461ms
finish merge:  78.05017ms
finish merge:  90.522502ms
finish merge:  122.553175ms
finished total: 122.748786ms

# ./test-linux-amd64 
godror WARNING: discrepancy between DBTIMEZONE ("+00:00"=0) and SYSTIMESTAMP ("+08:00"=800) - set connection timezone, see https://github.com/godror/godror/blob/master/doc/timezone.md
3: start to merge
0: start to merge
2: start to merge
l:  24 bindall:  10
l:  24 bindall:  10
1: start to merge
l:  24 bindall:  10
l:  24 bindall:  10
finish merge:  42.144107ms
can't merge:  ORA-00060: deadlock detected while waiting for resource
can't merge:  context deadline exceeded
can't merge:  context deadline exceeded
finished total: 5.257030257s
```

**Comment by freeNestor** [comment]
I would advise bulk merge return the first actual ora error like godror, not "ORA-24381: error(s) in array DML\n" which is not helpful to analysis.

**Comment by sijms** [comment]
very good now I have model to follow
I will continue investigation until issue resolved

**Comment by sijms** [comment]
fixed in v2.7.4

**Comment by freeNestor** [comment]
> fixed in v2.7.4

sorry, test result shows it improve, but timeout duration not expected, and I checked trace.log that error was "Connection Break With Error: read tcp 192.168.x.x:40870->192.168.x.x:1521: i/o timeout", so if timeout it will force close connection? I guess this behavior cause database session not exit normally or rollback and locks are not released, because error "ORA-00054: resource busy and acquire with NOWAIT specified or timeout expired" always occurs when truncating table after test program terminated, this means dead locks are not released and oracle will release it in 60s default. This situation is dangerous for production.

```
# ./test
3: start to merge
1: start to merge
2: start to merge
0: start to merge
can't merge:  read tcp 192.168.x.15:40864->192.168.x.47:1521: i/o timeout
can't merge:  read tcp 192.168.x.15:40866->192.168.x.47:1521: i/o timeout
can't merge:  read tcp 192.168.x.15:40868->192.168.x.47:1521: i/o timeout
can't merge:  read tcp 192.168.x.15:40870->192.168.x.47:1521: i/o timeout
finished total: 10.010796086s
```

can you make the request canceled like other driver and this will release dead lock in time, not close connection which not good for production.
other driver debug shows it cancel the operation.

```
can't merge:  context deadline exceeded
ODPI [01115] 2023-05-08 14:17:14.874: OCI error ORA-01013: user requested cancel of current operation (dpiStmt_executeMany / execute)
```

**Comment by sijms** [comment]
OK when I run only one go-routine with timeout the result as expected
```golang
0: start to merge
can't merge:  ORA-01013: user requested cancel of current operation
```
now see what happen
```sql
select count(*) from testshort

--result
-- 311
```
this means operation is not cancelled completely
to avoid this you should use transactions. after that
```sql
select count(*) from testshort

--result
-- 0
```
all the above with one go-routine
but when using multiple go-routines as I explain to you above it will produce deadlock and server will not respond in normal way so this is the result
```
3: start to merge
0: start to merge
1: start to merge
2: start to merge
finish merge:  41.33264ms
finish merge:  66.951215ms
can't merge:  context deadline exceeded
error in rollback:  read tcp 127.0.0.1:53464->127.0.0.1:1521: i/o timeout
can't merge:  write tcp 127.0.0.1:53464->127.0.0.1:1521: i/o timeout
finished total: 1.109113401s
```

as you see write and read operation are blocked so `merge and rollback` return timeout
result unexpected
```sql
select count(*) from testshort

--result
-- 379
```
so running this kind of sql in parallel is not recommended

**Comment by freeNestor** [comment]
> but when using multiple go-routines as I explain to you above it will produce deadlock and server will not respond in normal way so this is the result

I set context timeout to 5s, why my test result shows duration greater than 10s?

> so running this kind of sql in parallel is not recommended

Our app will split primary keys to different go-routines to ensure merge not conflict actually in production, so dead lock will never happen normally. I just want to say there will be many app running on the same database, we cannot ensure other app and ours won't cause dead lock in some circumstances we could not take into consideration. So we expect database operation timeout in our control and release locks after timeout in time, not in this test scenario only.
By the way, in my last test, dead locks are not released even test program exited.

> to avoid this you should use transactions
I know this, we don't care about transaction consistent in bulk merge this scenario, we care about timeout and database locks releases.

> 1- we have 4 go-routines which is started almost at same time
2- Exec operation is started with session.Write() which will take long time as we send 500 to 1000 items so server will keep
receiving data till finish
3- while server is receiving data another session.Write() commands started.
4- Now the server receive multiple merge requests at same time which make a dead lock inside the server and you can realize that at step 7
5- read operation now need to take place and because of dead lock it will stuck
6- read timeout occurs => break connection which will make write then read which also will stuck
7- it you wait for time you will receive error server dead lock occur

at 6th step, couldn't you use "session.Cancel()" something like this to cancel database operation? In my test, break connection make database session not closed normally and locks exist for long time. It's different compared with godror.
I'm appreciate for your work and efforts finally anyway.

**Comment by sijms** [comment]
`BreakConnection` it name due to its function which is send break message then reset message this is the way oracle drivers
cancel the current operation and receiving errors so it means cancel

**Comment by sijms** [comment]
I find something important
this is the implementation of BreakConnection/Cancel in original drivers
C#
```c#
internal void Break()
{
  if (this.m_sessionCtx.m_onBreakReset)
    return;
  this.m_sessionCtx.m_onBreakReset = true;
  if ((this.m_sessionCtx.m_negotiatedOptions & (int) TNSPacketOffsets.NSGRECVATTN /*1024*/) > 0)
    this.m_sessionCtx.m_transportAdapter.SendUrgent(new byte[1]
    {
      (byte) 33
    }, 0, 1);
  else
    this.SendMarker((int) MarkerPacket.NIQBMARK);
}
```
python
```python
# if we can use OOB, send an urgent message now followed by a reset
# marker to see if the server understands it
if self._caps.supports_oob \
        and self._caps.protocol_version >= TNS_VERSION_MIN_OOB_CHECK:
        self._socket.send(b"!", socket.MSG_OOB)
        self._send_marker(self._write_buf, TNS_MARKER_TYPE_RESET)
```

now i want to send out of bound message `MSG_OOB` to the server using golang
if you have any suggestion?

**Comment by sijms** [comment]
I success in sending `MSG_OOB`
now this code 
```golang
execCtx, execCancel := context.WithTimeout(context.Background(), 1*time.Second)
defer execCancel()
_, err = conn.ExecContext(execCtx, "begin DBMS_LOCK.sleep(5); end;")
if err != nil {
	fmt.Println(err)
	return
}
```
before I receive message 
```
ORA-01013: user requested cancel of current operation
```
 after 5 second = after finish sleep
now I receive message
```
ORA-01013: user requested cancel of current operation
ORA-06512: at "SYS.DBMS_LOCK", line 215
ORA-06512: at line 1
```
 after exactly 1 second

now I am testing our issue

**Comment by freeNestor** [comment]
> I success in sending MSG_OOB

Good news. Sorry, I'm not familiar with lower driver coding, I've searched MSG_OOB and oracle OSqlStmt C++ libraries Cancel() method, I think they use OOB mechanism to achieve this.

**Comment by sijms** [comment]
after fixing `MSG_OOB` now there is some improvement but not 100% 
if I test with autonomous database (through internet) timeout occur but no dead lock
if I test on local machine oracle docker sometimes server dead lock occur which will lead to resource lock with NOWAIT error

also note now there are 2 timeout one used by context which is define timeout for the operation itself so when time
finish and operation not done client will request cancel for current operation. during cancel global timeout will be used
which is defined by url option `timeout=integer` default is 30 seconds. I explain this in `readme` file `v2.7.4`

**Comment by freeNestor** [comment]
I tested and timeout is improved as expected.
```
·> ./test
3: start to merge
0: start to merge
2: start to merge
1: start to merge
finish merge:  35.853875ms
can't merge:  read tcp 10.1.67.114:55194->192.168.97.47:1521: i/o timeout
can't merge:  read tcp 10.1.67.114:55195->192.168.97.47:1521: i/o timeout
can't merge:  read tcp 10.1.67.114:55193->192.168.97.47:1521: i/o timeout
finished total: 5.001985958s
```

I analysis through netstat command on client and server:
client show two "FIN_WAIT_2" states after text program exited:
```
·> netstat -an|grep 1521
tcp4       0      0  x.xy.67.114.59050      192.168.x.xy.1521     FIN_WAIT_2
tcp4       0      0  x.xy.67.114.59049      192.168.x.xy.1521     FIN_WAIT_2
tcp4       0      0  x.xy.67.114.59047      192.168.x.xy.1521     TIME_WAIT
tcp4       0      0  x.xy.67.114.59048      192.168.x.xy.1521     TIME_WAIT
```
so "FIN_WAIT_2" means client has closed tcp socket and received ACK from server successfully, but wait for FIN from server. oracle server couldn't send FIN because it waits merge sql return data, until oracle force killed process who holding the lock. It seems MSG_OOB does not successfully cancelled the operation actually. If successfully cancelled, it will return data and send FIN to client which will turn tcp state to TIME_WAIT.

**Comment by sijms** [comment]
to cancel the operation you should send connection break followed by connection reset
connection break done by OOB by connection reset will make timeout because server is not responding to the client request

**Comment by freeNestor** [comment]
OK. Is this as far as we could go now? I'm curious about how other driver achieve this(when context timeout, there's no FIN_WAIT_2 state and locks after test program exited), I digged their source code found no clue.

```
# ./test-linux-amd64 godror 4
3: start to merge
0: start to merge
1: start to merge
2: start to merge
godror WARNING: discrepancy between DBTIMEZONE ("+00:00"=0) and SYSTIMESTAMP ("+08:00"=800) - set connection timezone, see https://github.com/godror/godror/blob/master/doc/timezone.md
finish merge:  181.507755ms
can't merge:  ORA-00060: deadlock detected while waiting for resource
can't merge:  context deadline exceeded
can't merge:  context deadline exceeded
finished total: 5.272152512s
# netstat -antp|grep 1521
tcp        0      0 192.168....:37433    192.168....:1521      TIME_WAIT   -                   
tcp        0      0 192.168....:37435    192.168....:1521      TIME_WAIT   -                   
tcp        0      0 192.168....:37397    192.168....:1521      TIME_WAIT   -                   
tcp        0      0 192.168....:37375    192.168....:1521      TIME_WAIT   -                   
tcp        0      0 192.168....:37346    192.168....:1521      TIME_WAIT   -
```

**Comment by sijms** [comment]
may be oracle original driver use something like mutex to avoid passing multiple long bulk merge request at the same time
as i tell you before if the first request make complete write before 2nd request start no deadlock will happen
but if 2nd request start write before first request finish write => deadlock will occur

**Comment by sijms** [comment]
also you can use 
```sql
_, err = tx.Exec(`LOCK TABLE TESTSHORT NOWAIT`)
if err != nil {
    return err
}
```
before merge and no deadlock will happen
